File: src/main/java/xf/xflp/opt/construction/strategy/HighestLowerLeft.java
package xf.xflp.opt.construction.strategy;

import xf.xflp.base.container.Container;
import xf.xflp.base.item.Item;
import xf.xflp.base.item.Position;
import xf.xflp.base.position.PositionCandidate;
import xf.xflp.exception.XFLPException;
import xf.xflp.exception.XFLPExceptionType;

import java.util.List;

/** 
 * Copyright (c) 2012-2022 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @author hschneid
 *
 * The strategy is used in construction heuristic to choose best possible insert position.
 *
 * This type of strategy chooses with maximal priority the highest and secondary the
 * most left (width) and most decent (length) position.
 *
 */
public class HighestLowerLeft extends BaseStrategy {

	@Override
	public PositionCandidate choose(Item item, Container container, List<PositionCandidate> candidates) throws XFLPException {
		if(candidates == null || candidates.isEmpty()) {
			throw new XFLPException(XFLPExceptionType.ILLEGAL_STATE, "List of positions must be not empty or null.");
		}

		List<PositionCandidate> filteredPositions = getPositionWithMinValue(
				candidates,
				this::getDistanceZ
		);

		if(filteredPositions.size() > 1) {
			filteredPositions = getPositionWithMinValue(
					filteredPositions,
					this::getDistance
			);
		}

		if(filteredPositions.isEmpty()) {
			throw new XFLPException(XFLPExceptionType.ILLEGAL_STATE, "There must be at least one position.");
		}

		return filteredPositions.get(0);
	}

	float getDistance(PositionCandidate candidate) {
		if(candidate == null) {
			return Float.MAX_VALUE;
		}

		Position p = candidate.position;
		return (float)Math.pow(
				(p.getX() * p.getX()) +
						(p.getY() * p.getY()) +
						(p.getZ() * p.getZ()), 0.5
		);
	}

	float getDistanceZ(PositionCandidate p) {
		return p == null ? Float.MAX_VALUE : (float) p.position.z * -1;
	}
}

File: src/xflp/opt/construction/strategy/highest_lower_left.py
# Copyright (c) 2012-2025 Holger Schneider
# All rights reserved.
#
# This source code is licensed under the MIT License (MIT) found in the
# LICENSE file in the root directory of this source tree.

"""
The strategy is used in construction heuristic to choose best possible insert position.

This type of strategy chooses with maximal priority the highest and secondary the
most left (width) and most decent (length) position.

Author: hschneid
"""

from typing import List, TYPE_CHECKING
import math

from xflp.opt.construction.strategy.base_strategy import BaseStrategy
from xflp.exception.xflp_exception import XFLPException, XFLPExceptionType

if TYPE_CHECKING:
    from xflp.base.container.container import Container
    from xflp.base.item.item import Item
    from xflp.base.position.position_candidate import PositionCandidate


class HighestLowerLeft(BaseStrategy):
    """
    Strategy that prioritizes highest positions, then lowest left positions.
    """

    def choose(
        self,
        item: 'Item',
        container: 'Container',
        candidates: List['PositionCandidate']
    ) -> 'PositionCandidate':
        """
        Choose the highest, then lower-left position.

        Args:
            item: The item to be placed
            container: The container
            candidates: List of position candidates

        Returns:
            The chosen position candidate

        Raises:
            XFLPException: If candidates list is empty or None
        """
        if candidates is None or len(candidates) == 0:
            raise XFLPException(
                XFLPExceptionType.ILLEGAL_STATE,
                "List of positions must be not empty or null."
            )

        # First, filter by minimum Z distance (highest position)
        filtered_positions = self.get_position_with_min_value(
            candidates,
            self._get_distance_z
        )

        # If multiple positions at same height, use 3D distance
        if len(filtered_positions) > 1:
            filtered_positions = self.get_position_with_min_value(
                filtered_positions,
                self._get_distance
            )

        if len(filtered_positions) == 0:
            raise XFLPException(
                XFLPExceptionType.ILLEGAL_STATE,
                "There must be at least one position."
            )

        return filtered_positions[0]

    def _get_distance(self, candidate: 'PositionCandidate') -> float:
        """
        Calculate 3D Euclidean distance from origin.

        Args:
            candidate: Position candidate

        Returns:
            Distance value
        """
        if candidate is None:
            return float('inf')

        p = candidate.position
        return math.sqrt(p.x ** 2 + p.y ** 2 + p.z ** 2)

    def _get_distance_z(self, candidate: 'PositionCandidate') -> float:
        """
        Get negative Z coordinate (for minimization to find highest position).

        Args:
            candidate: Position candidate

        Returns:
            Negative Z coordinate
        """
        if candidate is None:
            return float('inf')
        return -float(candidate.position.z)
